meta-client/src/main/java/org/apache/helix/metaclient/exception/MetaClientException.java

```diff
@@ -36,4 +36,168 @@ public MetaClientException(Throwable cause) {
+        return "Inconsistency was found.";
+      }
+    },
+
+    /** Session is moved or expired or non-exist. */
+    SESSION_ERROR {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Session is moved or expired or non-exist.";
+      }
+    },
+
+    /** Indicates a system and DB client or a usage errors not defined by following codes.
+     * It also indicate a range. Any value larger than this indicating error from
+     * client or caused by wrong usage.
+     */
+    DB_USER_ERROR {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Client or usage error.";
+      }
+    },
+
+    /** The listener does not exists. */
+    INVALID_LISTENER {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Listener does not exists.";
+      }
+    },
+
+    /** Authentication failed. */
+    AUTH_FAILED {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "authentication failed";
+      }
+    },
+
+    /** Invalid arguments. */
+    INVALID_ARGUMENTS {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Invalid arguments";
+      }
+    },
+
+    /** Version conflict. Return when caller tries to edit an entry with a specific version but
+     * the actual version of the entry on server is different. */
+    BAD_VERSION {
+      @Override
+      public MetaClientException createMetaClientExceptionFromReturnCode(ReturnCode returnCode) {
+        return new MetaClientBadVersionException();
+      }
+
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return " Version conflict.";
+      }
+    },
+
+    /** Entry already exists. Return when try to create a duplicated entry. */
+    ENTRY_EXISTS {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Entry already exists.";
+      }
+    },
+
+    /** The client is not Authenticated. */
+    NO_AUTH {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Not Authenticated.";
+      }
+    },
+
+    /** Entry does not exist. */
+    NO_SUCH_ENTRY {
+      @Override
+      public MetaClientException createMetaClientExceptionFromReturnCode(ReturnCode returnCode) {
+        return new MetaClientNoNodeException();
+      }
+
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "Entry does not exist.";
+      }
+    },
+
+    /**The entry has sub entries. Return when operation can only be down at entry with no
+     * sub entries. (i.e. unrecursively delete an entry . )*/
+    NOT_LEAF_ENTRY {
+      @Override
+      public String getCodeMessage(ReturnCode returnCode) {
+        return "The entry has sub entries.";
+      }
+    };
+
+    public MetaClientException createMetaClientExceptionFromReturnCode(ReturnCode returnCode) {
```

Comment:
The input param can be dropped.
Also nit on naming, since this is applied on return code already, ...FromReturnCode is redundant.